### PR TITLE
Add DevServices for Mongo

### PR DESCRIPTION
**Pull request by geoand**
~~This still needs more work:~~

- [x] Tie it in with `QuarkusIntegrationTest` and `NativeImageTest`
- [x] Convert some of the tests to use it
- [x] Add a note in the documentation

But I wanted to get early feedback from @stuartwdouglas on the approach. Basically this is a watered down version of the DevServices for Datasources as we don't need any of the complexities of Datasource handling

### CODE REVIEW

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientBuildTimeConfig.java [hunk 1]
```diff
@@ -29,4 +29,10 @@
      */
     @ConfigItem(name = "force-default-clients")
     public boolean forceDefaultClients;
+
+    /**
+     * Configuration for DevServices. DevServices allows Quarkus to automatically start MongoDB in dev and test mode.
+     */
+    @ConfigItem
+    public DevServicesBuildTimeConfig devServices;
```
Comment (geoand) [inline extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientBuildTimeConfig.java:37]: For now I have not enabled dev services for the named connections - just the default one

**Review by stuartwdouglas — COMMENTED** [review]
I think this looks good. I do wonder if there is any common parts that we could refactor into one processor, but I am not sure if it is worth the effort.

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/DevServicesMongoProcessor.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        if (!needToStart) {
+            // a connection string has been provided
+            log.debug("Not starting devservices for " + (isDefault(connectionName) ? "default datasource" : connectionName)
+                    + " as a connection string has been provided");
+            return null;
+        }
+
+        MongoDBContainer mongoDBContainer;
+        if (capturedProperties.imageName != null) {
+            mongoDBContainer = new MongoDBContainer(DockerImageName.parse(capturedProperties.imageName));
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/DevServicesMongoProcessor.java:148]: Oh, that's nice, thanks!

**Comment by geoand** [comment]
> I think this looks good. I do wonder if there is any common parts that we could refactor into one processor, but I am not sure if it is worth the effort.

I was wondering the same, but given that Datasource handling is much more complex, I am leaning towards that it's not worth the effort

**Comment by geoand** [comment]
PR is ready to go.

In this incarnation, only the default connection string is supported, but the infrastructure to support named connections string in the future

**Comment by geoand** [comment]
> > I think this looks good. I do wonder if there is any common parts that we could refactor into one processor, but I am not sure if it is worth the effort.
> 
> I was wondering the same, but given that Datasource handling is much more complex, I am leaning towards that it's not worth the effort

It might make sense to try and extract some common stuff when the next service is added (like Redis or something) as the operations will be almost identical to what I did here for Mongo

**Comment by geoand** [comment]
CI failures should be fixed now

**Comment by geoand** [comment]
@stuartwdouglas is this okay with you?

**Review by stuartwdouglas — APPROVED** [review]

**Pull request merged**